In [12]:
import json
from pprint import pprint
with open("空氣品質aqi.json",encoding="utf-8") as file:
    data:dict = json.load(file)
contents:list[dict]= data['records']
pprint(contents)

[{'aqi': '64',
  'co': '0.18',
  'co_8hr': '0.1',
  'county': '屏東縣',
  'datacreationdate': '2024-03-15 09:00',
  'latitude': '22.260899',
  'longitude': '120.651472',
  'no': '0.3',
  'no2': '2.1',
  'nox': '2.4',
  'o3': '58.1',
  'o3_8hr': '60.0',
  'pm10': '25',
  'pm10_avg': '25',
  'pm2.5': '14',
  'pm2.5_avg': '12.5',
  'pollutant': '臭氧八小時',
  'siteid': '313',
  'sitename': '屏東(枋山)',
  'so2': '0.4',
  'so2_avg': '0',
  'status': '普通',
  'unit': '',
  'winddirec': '91',
  'windspeed': '7.3'},
 {'aqi': '78',
  'co': '0.76',
  'co_8hr': '0.4',
  'county': '新北市',
  'datacreationdate': '2024-03-15 09:00',
  'latitude': '24.94902778',
  'longitude': '121.38352778',
  'no': '25.4',
  'no2': '24.5',
  'nox': '50',
  'o3': '5.4',
  'o3_8hr': '9.8',
  'pm10': '44',
  'pm10_avg': '37',
  'pm2.5': '27',
  'pm2.5_avg': '26.6',
  'pollutant': '細懸浮微粒',
  'siteid': '311',
  'sitename': '新北(樹林)',
  'so2': '0.8',
  'so2_avg': '0',
  'status': '普通',
  'unit': '',
  'winddirec': '109',
  'windspeed'

[{'aqi': '64',
  'co': '0.18',
  'co_8hr': '0.1',
  'county': '屏東縣',
  'datacreationdate': '2024-03-15 09:00',
  'latitude': '22.260899',
  'longitude': '120.651472',
  'no': '0.3',
  'no2': '2.1',
  'nox': '2.4',
  'o3': '58.1',
  'o3_8hr': '60.0',
  'pm10': '25',
  'pm10_avg': '25',
  'pm2.5': '14',
  'pm2.5_avg': '12.5',
  'pollutant': '臭氧八小時',
  'siteid': '313',
  'sitename': '屏東(枋山)',
  'so2': '0.4',
  'so2_avg': '0',
  'status': '普通',
  'unit': '',
  'winddirec': '91'
  'windspeed': '7.3'},

In [13]:
from pydantic import BaseModel,Field,field_validator
from datetime import datetime

class AirSite(BaseModel):
    aqi:int
    county:str
    date:datetime = Field(alias="datacreationdate")
    lat:float = Field(alias = "latitude")
    lon:float = Field(alias="longitude")
    pm25:float = Field(alias="pm2.5")
    pollutant:str 
    site_name:str = Field(alias="sitename")
    status:str  

    @field_validator('aqi','lat','lon','pm25', mode='before')
    @classmethod
    def empty_to_zero(cls, v):
        return 0 if v == '' else v


In [14]:
class Root(BaseModel):
    status:bool = True
    sites:list[AirSite]

In [15]:
root = Root(sites=[AirSite(**item) for item in contents])

In [17]:
for site in root.sites:
    print(site)

aqi=64 county='屏東縣' date=datetime.datetime(2024, 3, 15, 9, 0) lat=22.260899 lon=120.651472 pm25=14.0 pollutant='臭氧八小時' site_name='屏東(枋山)' status='普通'
aqi=78 county='新北市' date=datetime.datetime(2024, 3, 15, 9, 0) lat=24.94902778 lon=121.38352778 pm25=27.0 pollutant='細懸浮微粒' site_name='新北(樹林)' status='普通'
aqi=94 county='屏東縣' date=datetime.datetime(2024, 3, 15, 9, 0) lat=22.35222 lon=120.37722 pm25=33.0 pollutant='細懸浮微粒' site_name='屏東(琉球)' status='普通'
aqi=152 county='臺南市' date=datetime.datetime(2024, 3, 15, 9, 0) lat=23.17904722 lon=120.24583056 pm25=61.0 pollutant='細懸浮微粒' site_name='臺南(麻豆)' status='對所有族群不健康'
aqi=148 county='高雄市' date=datetime.datetime(2024, 3, 15, 9, 0) lat=22.87985556 lon=120.24535 pm25=57.0 pollutant='細懸浮微粒' site_name='高雄(湖內)' status='對敏感族群不健康'
aqi=119 county='彰化縣' date=datetime.datetime(2024, 3, 15, 9, 0) lat=23.96117 lon=120.56373 pm25=41.0 pollutant='細懸浮微粒' site_name='彰化(員林)' status='對敏感族群不健康'
aqi=84 county='彰化縣' date=datetime.datetime(2024, 3, 15, 9, 0) lat=23.85493